infrastructure/comms/src/peer_manager/node_id.rs

```diff
@@ -0,0 +1,252 @@
+pub struct NodeDistance([u8; 32]);
+
+impl NodeDistance {
+    /// Construct a new zero distance
+    pub fn new() -> NodeDistance {
+        (NodeDistance::from(&[0; 32]))
+    }
+
+    /// Construct a node distance from an array of 32 bytes
+    pub fn from(array: &[u8; 32]) -> NodeDistance {
+        (NodeDistance(*array))
+    }
+}
+
+impl PartialEq for NodeDistance {
+    fn eq(&self, nd: &NodeDistance) -> bool {
+        for i in 0..self.0.len() {
+            if self.0[i] != nd.0[i] {
+                return false;
+            }
+        }
+        (true)
+    }
+}
+
+#[derive(Clone, Debug, Eq)]
+pub struct NodeID([u8; 32]);
+
+impl NodeID {
+    /// Construct a new node id on the origin
+    pub fn new() -> NodeID {
+        (NodeID::from(&[0; 32]))
+    }
+
+    /// Construct a node id from an array of 32 bytes
+    pub fn from(array: &[u8; 32]) -> NodeID {
+        (NodeID(*array))
+    }
+
+    /// Derive a node id from a public key: node_id=hash(public_key)
+    pub fn from_public_key<K: PublicKey, D: Digest>(public_key: &K) -> NodeID {
+        let mut array = [0; 32];
+        array.copy_from_slice(D::digest(&public_key.as_bytes()).as_slice());
+        (NodeID(array))
+    }
+
+    /// Generate a node id from a base layer registration using the block hash and public key
+    // pub fn from_baselayer_registration<D: Digest>(......) -> NodeID {
+    // TODO: NodeId=hash(blockhash(with block_height),public key?)
+    // }
+
+    /// Calculate the distance between the current node id and the provided node id using the XOR metric
+    pub fn distance(&self, node_id: &NodeID) -> NodeDistance {
+        (xor_metric(&self, &node_id))
+    }
+
+    /// Find and return the indices of the K nearest neighbours from the provided node id list
+    pub fn closest_indices(&self, node_ids: &Vec<NodeID>, k: usize) -> Result<Vec<usize>, NodeIDError> {
+        if k > node_ids.len() {
+            return Err(NodeIDError::OutOfBounds);
+        }
+        let mut indices: Vec<usize> = Vec::with_capacity(node_ids.len());
+        let mut dists: Vec<NodeDistance> = Vec::with_capacity(node_ids.len());
+        for i in 0..node_ids.len() {
+            indices.push(i);
+            dists.push(self.distance(&node_ids[i]));
+        }
+        // Perform partial sort of elements only up to K elements
+        let mut nearest_node_indices: Vec<usize> = Vec::with_capacity(k);
+        for i in 0..k {
+            for j in i + 1..node_ids.len() {
+                if dists[i] > dists[j] {
+                    dists.swap(i, j);
+                    indices.swap(i, j);
+                }
+            }
+            nearest_node_indices.push(indices[i]);
+        }
+        Ok(nearest_node_indices)
+    }
+
+    /// Find and return the node ids of the K nearest neighbours from the provided node id list
+    pub fn closest(&self, node_ids: &Vec<NodeID>, k: usize) -> Result<Vec<NodeID>, NodeIDError> {
+        match self.closest_indices(&node_ids, k) {
```

Comment:
Think use `?` here:

```rust
let closest_node_indices = self.closest_indices(&node_ids, k)?;
```

Reply:
In this case the match is used to safely unwrap the indices to compile node id list.

Reply:
Using `?` means you don't have to include the superfluous `Err(e) => Err(e)` part at the bottom. 

The entire function body would be:

```rust
        let nearest_node_indices = self.closest_indices(&node_ids, k)?;
        let mut nearest_node_ids: Vec<NodeId> = Vec::with_capacity(nearest_node_indices.len());
        for i in 0..nearest_node_indices.len() {
            nearest_node_ids.push(node_ids[nearest_node_indices[i]].clone());
        }
        Ok(nearest_node_ids)
```

Exactly the same as your code, but more concise

Reply:
Or even
```rust
let nearest_node_indices = self.closest_indices(&node_ids, k)?;
        let mut nearest_node_ids: Vec<NodeId> = Vec::with_capacity(nearest_node_indices.len());
        for nearest in nearest_node_indices.iter() {
            nearest_node_ids.push(node_ids[nearest].clone());
        }
        Ok(nearest_node_ids)
```

Reply:
Nice, that looks good!